pkg/app/piped/trigger/determiner.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trigger
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/filematcher"
+	"github.com/pipe-cd/pipe/pkg/git"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type LastTriggeredCommitGetter interface {
+	Get(ctx context.Context, applicationID string) (string, error)
+}
+
+type Determiner struct {
+	repo         git.Repo
+	targetCommit string
+	commitGetter LastTriggeredCommitGetter
+	logger       *zap.Logger
+}
+
+func NewDeterminer(repo git.Repo, targetCommit string, cg LastTriggeredCommitGetter, logger *zap.Logger) *Determiner {
+	return &Determiner{
+		repo:         repo,
+		targetCommit: targetCommit,
+		commitGetter: cg,
+		logger:       logger.Named("determiner"),
+	}
+}
+
+// ShouldTrigger decides whether a given application should be triggered or not.
+func (d *Determiner) ShouldTrigger(ctx context.Context, app *model.Application) (bool, error) {
+	logger := d.logger.With(
+		zap.String("app", app.Name),
+		zap.String("app-id", app.Id),
+		zap.String("target-commit", d.targetCommit),
+	)
+
+	preCommit, err := d.commitGetter.Get(ctx, app.Id)
+	if err != nil {
+		logger.Error("failed to get last triggered commit", zap.Error(err))
+		return false, err
+	}
+	if preCommit == "" {
+		logger.Info("no previously triggered deployment was found")
```

Comment:
Isn't it fine to return `true` at this time? I feel like it should trigger whatever `targetCommit` is.

Reply:
Nice catch!